src/System.Private.CoreLib/shared/System/String.Comparison.cs

```diff
@@ -746,13 +746,21 @@ public static bool Equals(string a, string b, StringComparison comparisonType)
         // they will return the same hash code.
         public override int GetHashCode()
         {
-            return Marvin.ComputeHash32(ref Unsafe.As<char, byte>(ref _firstChar), _stringLength * 2, Marvin.DefaultSeed);
+            ulong seed = Marvin.DefaultSeed;
```

Comment:
Does this affect performance the regular string GetHashCode?

Reply:
I didn't see any regression. However, due to https://github.com/dotnet/coreclr/issues/20245, retrieval of the "default seed" field is going through the typical AppDomain static lookup logic rather than as a constant burned directly in to the call site. I wouldn't put much faith in any collected numbers for regular `GetHashCode` perf runs until that issue is resolved.

Reply:
> retrieval of the "default seed" field is going through the typical AppDomain static lookup logic

It does this only if you got a loop in a method that never returns. If you change the test to be a more realistic pattern with method that returns that lets tiered JIT to kick in, the default seed gets inlined perfectly in steady state. E.g. Try: 


```csharp
using System;

class Test
{
   static int DoIt(string s)
   {
       int r = s.Length;
       for (int i = 0; i < 1000; i++) r += s.GetHashCode();
       return r;
   }

   static void Main()
   {
       for (;;) { DoIt("Hello"); }
   }
}
```

The code for `DoIt` in steady state is as good as it can be wrt. Seed:

```
00007ffa`7e641760 57              push    rdi
00007ffa`7e641761 56              push    rsi
00007ffa`7e641762 55              push    rbp
00007ffa`7e641763 53              push    rbx
00007ffa`7e641764 4883ec28        sub     rsp,28h
00007ffa`7e641768 488bf1          mov     rsi,rcx
00007ffa`7e64176b 8b7e08          mov     edi,dword ptr [rsi+8]
00007ffa`7e64176e 33db            xor     ebx,ebx
00007ffa`7e641770 488d6e0c        lea     rbp,[rsi+0Ch]
00007ffa`7e641774 488bcd          mov     rcx,rbp
00007ffa`7e641777 8b5608          mov     edx,dword ptr [rsi+8]
00007ffa`7e64177a d1e2            shl     edx,1
00007ffa`7e64177c 49b8f2c1ed8df3784748 mov r8,484778F38DEDC1F2h <- seed 
00007ffa`7e641786 e87561d05e      call    System_Private_CoreLib!System.Marvin.ComputeHash32(Byte ByRef, Int32, UInt64) (00007ffa`dd347900)
00007ffa`7e64178b 03f8            add     edi,eax
00007ffa`7e64178d ffc3            inc     ebx
00007ffa`7e64178f 81fbe8030000    cmp     ebx,3E8h
00007ffa`7e641795 7cdd            jl      test!Test.DoStuff(System.String)+0x94 (00007ffa`7e641774)
00007ffa`7e641797 8bc7            mov     eax,edi
00007ffa`7e641799 4883c428        add     rsp,28h
00007ffa`7e64179d 5b              pop     rbx
00007ffa`7e64179e 5d              pop     rbp
00007ffa`7e64179f 5e              pop     rsi
00007ffa`7e6417a0 5f              pop     rdi
00007ffa`7e6417a1 c3              ret
```

Reply:
@jkotas This is what my inner loop looks like using the exact code you provided.

```asm
00007ff8`9660176f 488bce          mov     rcx,rsi
00007ff8`96601772 e8693fe855      call    System_Private_CoreLib!System.String.GetHashCode() (00007ff8`ec4856e0)
00007ff8`96601777 03f8            add     edi,eax
00007ff8`96601779 ffc3            inc     ebx
00007ff8`9660177b 81fbe8030000    cmp     ebx,3E8h
00007ff8`96601781 7cec            jl      ConsoleAppPrivateRuntime!Program.DoIt(System.String)+0x8f (00007ff8`9660176f)
```

Note the non-inlined call to `String.GetHashCode`. I'm using 3.0.0-preview1-27008-04 for this run, which I believe introduced a regression here, because I believe that you're right in that the sample code _should_ be fully optimized.

With `[MethodImpl((MethodImplOptions)0x0200)]` on the `DoIt` method I get the correct codegen for the inner loop:

```asm
00007ff8`96615954 488bcd          mov     rcx,rbp
00007ff8`96615957 8b5608          mov     edx,dword ptr [rsi+8]
00007ff8`9661595a d1e2            shl     edx,1
00007ff8`9661595c 49b873c833f899c65d8e mov r8,8E5DC699F833C873h
00007ff8`96615966 e8b5d2fb55      call    System_Private_CoreLib!System.Marvin.ComputeHash32(Byte ByRef, Int32, UInt64) (00007ff8`ec5d2c20)
00007ff8`9661596b 03f8            add     edi,eax
00007ff8`9661596d ffc3            inc     ebx
00007ff8`9661596f 81fbe8030000    cmp     ebx,3E8h
00007ff8`96615975 7cdd            jl      ConsoleAppPrivateRuntime!Program.DoIt(System.String)+0x14 (00007ff8`96615954)
```

Reply:
Anyway, long story short, the baseline and new runs for `String.GetHashCode` are within 0.5% of each other. I'm gonna call that a wash.

I did have to slap `[MethodImpl(MethodImplOptions.AggressiveInlining)]` on normal `String.GetHashCode` in the end due to the increased IL size. But TBH that seems fine to me. Do you have any concerns with this?

New inner loop:

```asm
00007ff8`965f19ed 498bce          mov     rcx,r14
00007ff8`965f19f0 8b5608          mov     edx,dword ptr [rsi+8]
00007ff8`965f19f3 d1e2            shl     edx,1
00007ff8`965f19f5 41b8cd3e0c49    mov     r8d,490C3ECDh
00007ff8`965f19fb 41b993e4fe70    mov     r9d,70FEE493h
00007ff8`965f1a01 e83a61ec55      call    System_Private_CoreLib!System.Marvin.ComputeHash32(Byte ByRef, Int32, UInt32, UInt32) (00007ff8`ec4b7b40)
00007ff8`965f1a06 03d8            add     ebx,eax
00007ff8`965f1a08 ffc5            inc     ebp
00007ff8`965f1a0a 81fd00e1f505    cmp     ebp,5F5E100h
00007ff8`965f1a10 7cdb            jl      ConsoleAppPrivateRuntime!Program.DoIt(System.String)+0x5d (00007ff8`965f19ed)
```

Reply:
Are you waiting for the benchmark to reach steady state before looking at the disassembly?

Note that you can no longer just set breakpoint and look at the disassembly with Tiered JITing.

Reply:
Yes, I let the benchmark run for 10 seconds or so before attaching windbg.

Reply:
Here's the reason code for why `String.GetHashCode` doesn't inline into the caller without the _AggressiveInlining_ marker:
`[0 IL=0019 TR=000025 060002CD] [FAILED: noinline per IL/cached result] String:GetHashCode():int:this`

And with the marker:
`[6 IL=0019 TR=000025 060002CD] [aggressive inline attribute] String:GetHashCode():int:this`